File: app/src/main/java/com/ywwynm/everythingdone/adapters/AudioAttachmentAdapter.java
package com.ywwynm.everythingdone.adapters;

import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.media.MediaPlayer;
import android.net.Uri;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.ywwynm.everythingdone.R;
import com.ywwynm.everythingdone.helpers.AttachmentHelper;
import com.ywwynm.everythingdone.utils.DateTimeUtil;
import com.ywwynm.everythingdone.utils.FileUtil;

import java.io.File;
import java.util.List;

/**
 * Created by ywwynm on 2015/10/4.
 * adapter for audio attachment
 */
public class AudioAttachmentAdapter extends RecyclerView.Adapter<AudioAttachmentAdapter.AudioCardViewHolder> {

    public static final String TAG = "AudioAttachmentAdapter";

    private Activity mActivity;

    private int mAccentColor;

    private boolean mEditable;

    private LayoutInflater mInflater;

    private List<String> mItems;

    private int mPlayingIndex = -1;
    private MediaPlayer mPlayer;

    private boolean mTakingScreenshot = false;

    public interface RemoveCallback {
        void onRemoved(int pos);
    }
    private RemoveCallback mRemoveCallback;

    public AudioAttachmentAdapter(
            Activity activity, int accentColor, boolean editable, List<String> items,
            RemoveCallback callback) {
        mActivity = activity;
        mAccentColor = accentColor;
        mEditable = editable;
        mInflater = LayoutInflater.from(activity);
        mItems = items;

        mRemoveCallback = callback;
    }

    public void setTakingScreenshot(boolean takingScreenshot) {
        mTakingScreenshot = takingScreenshot;
        if (mPlayer != null && mPlayer.isPlaying()) {
            stopPlaying();
        }
        notifyDataSetChanged();
    }

    public List<String> getItems() {
        return mItems;
    }

    public int getPlayingIndex() {
        return mPlayingIndex;
    }

    public void setPlayingIndex(int playingIndex) {
        mPlayingIndex = playingIndex;
    }

    @Override
    public AudioCardViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new AudioCardViewHolder(mInflater.inflate(R.layout.attachment_audio, parent, false));
    }

    @Override
    public void onBindViewHolder(AudioCardViewHolder holder, int position) {
        String typePathName = mItems.get(position);
        String pathName = typePathName.substring(1, typePathName.length());
        File file = new File(pathName);

        holder.tvName.setText(file.getName());
        long duration = FileUtil.getMediaDuration(pathName);
        holder.tvSize.setText(DateTimeUtil.getDurationBriefStr(duration));

        if (mTakingScreenshot) {
            holder.ivFirst .setVisibility(View.VISIBLE);
            holder.ivSecond.setVisibility(View.GONE);
            holder.ivThird .setVisibility(View.GONE);
            holder.ivFirst.setImageResource(R.drawable.act_play);
        } else {
            Context context = holder.itemView.getContext();
            holder.ivSecond.setVisibility(View.VISIBLE);
            holder.ivThird. setVisibility(View.VISIBLE);
            if (mPlayingIndex == position) {
                holder.ivFirst.setVisibility(View.VISIBLE);
                if (mPlayer.isPlaying()) {
                    holder.ivFirst.setImageResource(R.drawable.act_pause);
                    holder.ivFirst.setContentDescription(
                            context.getString(R.string.cd_pause_play_audio_attachment));
                } else {
                    holder.ivFirst.setImageResource(R.drawable.act_play);
                    holder.ivFirst.setContentDescription(
                            context.getString(R.string.cd_play_audio_attachment));
                }
                holder.ivSecond.setImageResource(R.drawable.act_stop_playing_audio);
                holder.ivSecond.setContentDescription(
                        context.getString(R.string.cd_stop_play_audio_attachment));
            } else {
                if (mEditable) {
                    holder.ivFirst.setVisibility(View.VISIBLE);
                    holder.ivFirst.setImageResource(R.drawable.act_play);
                    holder.ivFirst.setContentDescription(
                            context.getString(R.string.cd_play_audio_attachment));
                    holder.ivSecond.setImageResource(R.drawable.delete_audio);
                    holder.ivSecond.setContentDescription(
                            context.getString(R.string.cd_delete_audio_attachment));
                } else {
                    holder.ivFirst.setVisibility(View.GONE);
                    holder.ivSecond.setImageResource(R.drawable.act_play);
                    holder.ivSecond.setContentDescription(
                            context.getString(R.string.cd_play_audio_attachment));
                }
            }
        }
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    private void startPlaying(int index) {
        mPlayingIndex = index;
        String typePathName = mItems.get(index);
        File file = new File(typePathName.substring(1, typePathName.length()));

        mPlayer = MediaPlayer.create(mActivity, Uri.fromFile(file));
        mPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
            @Override
            public void onCompletion(MediaPlayer mp) {
                final int index1 = mPlayingIndex;
                stopPlaying();
                notifyItemChanged(index1);
            }
        });
        mPlayer.start();
    }

    public void stopPlaying() {
        mPlayingIndex = -1;
        mPlayer.stop();
        mPlayer.reset();
        mPlayer.release();
        mPlayer = null;
    }

    class AudioCardViewHolder extends BaseViewHolder {

        final CardView  cv;
        final TextView  tvName;
        final TextView  tvSize;
        final ImageView ivFirst;
        final ImageView ivSecond;
        final ImageView ivThird;

        AudioCardViewHolder(View itemView) {
            super(itemView);

            cv       = f(R.id.cv_audio_attachment);
            tvName   = f(R.id.tv_audio_file_name);
            tvSize   = f(R.id.tv_audio_size);
            ivFirst  = f(R.id.iv_card_audio_first);
            ivSecond = f(R.id.iv_card_audio_second);
            ivThird  = f(R.id.iv_card_audio_third);

            Drawable d = ContextCompat.getDrawable(
                    mActivity, R.drawable.act_show_attachment_info);
            Drawable d1 = d.mutate();
            d1.setColorFilter(Color.parseColor("#8A000000"), PorterDuff.Mode.SRC_ATOP);
            ivThird.setImageDrawable(d1);

            cv.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    togglePlay();
                }
            });

            ivThird.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    String item = mItems.get(AudioCardViewHolder.this.getAdapterPosition());
                    String pathName = item.substring(1, item.length());
                    AttachmentHelper.showAttachmentInfoDialog(mActivity, mAccentColor, pathName);
                }
            });

            if (mEditable) {
                setEventsEditable();
            } else {
                setEventsUneditable();
            }
        }

        private void setEventsEditable() {
            ivFirst.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    togglePlay();
                }
            });

            ivSecond.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int pos = getAdapterPosition();
                    if (mPlayingIndex == pos) {
                        stopPlaying();
                        notifyItemChanged(pos);
                    } else {
                        if (mRemoveCallback != null) {
                            mRemoveCallback.onRemoved(pos);
                        }
                    }
                }
            });
        }

        private void togglePlay() {
            int pos = getAdapterPosition();
            if (mPlayingIndex == pos) {
                if (mPlayer.isPlaying()) {
                    mPlayer.pause();
                } else {
                    mPlayer.start();
                }
            } else {
                if (mPlayingIndex != -1) {
                    final int index = mPlayingIndex;
                    stopPlaying();
                    notifyItemChanged(index);
                }
                startPlaying(pos);
            }
            notifyItemChanged(pos);
        }

        private void setEventsUneditable() {
            ivFirst.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mPlayer.isPlaying()) {
                        mPlayer.pause();
                    } else {
                        mPlayer.start();
                    }
                    notifyItemChanged(getAdapterPosition());
                }
            });

            ivSecond.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int pos = getAdapterPosition();
                    if (mPlayingIndex == pos) {
                         stopPlaying();
                    } else {
                        if (mPlayingIndex != -1) {
                            final int index = mPlayingIndex;
                            stopPlaying();
                            notifyItemChanged(index);
                        }
                        startPlaying(pos);
                    }
                    notifyItemChanged(pos);
                }
            });
        }
    }

}

File: app/src/main/java/com/ywwynm/everythingdone/adapters/AudioAttachmentAdapter.kt
@file:Suppress("DEPRECATION")

package com.ywwynm.everythingdone.adapters

import android.app.Activity
import android.media.MediaPlayer
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.ywwynm.everythingdone.R
import com.ywwynm.everythingdone.helpers.AttachmentHelper
import com.ywwynm.everythingdone.model.ThingBackground
import com.ywwynm.everythingdone.utils.DateTimeUtil
import com.ywwynm.everythingdone.utils.DisplayUtil
import com.ywwynm.everythingdone.utils.FileUtil
import com.ywwynm.everythingdone.views.GradientRippleDrawable

import java.io.File

/**
 * Created by ywwynm on 2015/10/4.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * adapter for audio attachment
 */
open class AudioAttachmentAdapter(
    activity: Activity?,
    accentColor: Int,
    editable: Boolean,
    items: List<String?>?,
    callback: RemoveCallback?
) : RecyclerView.Adapter<AudioAttachmentAdapter.AudioCardViewHolder>() {

    private var mActivity: Activity? = activity

    private var mAccentColor: Int = accentColor
    /** Phase 8: full accent so the attachment-info dialog launched from a tap
     *  can render gradient. Always wraps [mAccentColor] as PURE when
     *  unset, so the helper never receives null. */
    private var mAccentBackground: ThingBackground? = ThingBackground.pure(accentColor)

    private var mEditable: Boolean = editable

    private var mInflater: LayoutInflater? = LayoutInflater.from(activity)

    private var mItems: List<String?>? = items

    private var mPlayingIndex: Int = -1
    private var mPlayer: MediaPlayer? = null

    private var mTakingScreenshot: Boolean = false

    interface RemoveCallback {
        fun onRemoved(pos: Int)
    }
    private var mRemoveCallback: RemoveCallback? = callback

    /** Phase 8: upgrade the accent to a full [ThingBackground] so the
     *  attachment-info dialog renders gradient when the thing has one. */
    open fun setAccentBackground(bg: ThingBackground?) {
        if (bg == null) return
        mAccentBackground = bg
        mAccentColor = bg.representativeColor()
    }

    open fun setTakingScreenshot(takingScreenshot: Boolean) {
        mTakingScreenshot = takingScreenshot
        if (mPlayer != null && mPlayer!!.isPlaying) {
            stopPlaying()
        }
        notifyDataSetChanged()
    }

    open fun getItems(): List<String?>? = mItems

    open fun getPlayingIndex(): Int = mPlayingIndex

    open fun setPlayingIndex(playingIndex: Int) {
        mPlayingIndex = playingIndex
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AudioCardViewHolder {
        return AudioCardViewHolder(mInflater!!.inflate(R.layout.attachment_audio, parent, false))
    }

    override fun onBindViewHolder(holder: AudioCardViewHolder, position: Int) {
        val typePathName = mItems!![position]
        val pathName = typePathName!!.substring(1, typePathName.length)
        val file = File(pathName)

        holder.tvName!!.text = file.name
        val duration = FileUtil.getMediaDuration(pathName)
        holder.tvSize!!.text = DateTimeUtil.getDurationBriefStr(duration)
        holder.cv!!.setCardBackgroundColor(
            ContextCompat.getColor(mActivity!!, R.color.app_chrome_surface_elevated)
        )
        holder.tvName!!.setTextColor(
            ContextCompat.getColor(mActivity!!, R.color.app_chrome_on_surface_secondary)
        )
        holder.tvSize!!.setTextColor(
            ContextCompat.getColor(mActivity!!, R.color.app_chrome_on_surface_hint)
        )
        applyAudioRipples(holder)
        setAudioIcon(holder.ivThird, R.drawable.act_show_attachment_info)

        if (mTakingScreenshot) {
            holder.ivFirst !!.visibility = View.VISIBLE
            holder.ivSecond!!.visibility = View.GONE
            holder.ivThird !!.visibility = View.GONE
            setAudioIcon(holder.ivFirst, R.drawable.act_play)
        } else {
            val context = holder.itemView.context
            holder.ivSecond!!.visibility = View.VISIBLE
            holder.ivThird !!.visibility = View.VISIBLE
            if (mPlayingIndex == position) {
                holder.ivFirst!!.visibility = View.VISIBLE
                if (mPlayer!!.isPlaying) {
                    setAudioIcon(holder.ivFirst, R.drawable.act_pause)
                    holder.ivFirst!!.contentDescription =
                        context.getString(R.string.cd_pause_play_audio_attachment)
                } else {
                    setAudioIcon(holder.ivFirst, R.drawable.act_play)
                    holder.ivFirst!!.contentDescription =
                        context.getString(R.string.cd_play_audio_attachment)
                }
                setAudioIcon(holder.ivSecond, R.drawable.act_stop_playing_audio)
                holder.ivSecond!!.contentDescription =
                    context.getString(R.string.cd_stop_play_audio_attachment)
            } else {
                if (mEditable) {
                    holder.ivFirst!!.visibility = View.VISIBLE
                    setAudioIcon(holder.ivFirst, R.drawable.act_play)
                    holder.ivFirst!!.contentDescription =
                        context.getString(R.string.cd_play_audio_attachment)
                    setAudioIcon(holder.ivSecond, R.drawable.delete_audio)
                    holder.ivSecond!!.contentDescription =
                        context.getString(R.string.cd_delete_audio_attachment)
                } else {
                    holder.ivFirst!!.visibility = View.GONE
                    setAudioIcon(holder.ivSecond, R.drawable.act_play)
                    holder.ivSecond!!.contentDescription =
                        context.getString(R.string.cd_play_audio_attachment)
                }
            }
        }
    }

    override fun getItemCount(): Int = mItems!!.size

    override fun onViewRecycled(holder: AudioCardViewHolder) {
        super.onViewRecycled(holder)
        (holder.cv?.foreground as? GradientRippleDrawable)?.stopAnimations()
        (holder.ivFirst?.background as? GradientRippleDrawable)?.stopAnimations()
        (holder.ivSecond?.background as? GradientRippleDrawable)?.stopAnimations()
        (holder.ivThird?.background as? GradientRippleDrawable)?.stopAnimations()
    }

    private fun setAudioIcon(imageView: ImageView?, iconRes: Int) {
        imageView!!.setImageDrawable(
            DisplayUtil.opaqueTintDrawable(
                mActivity!!,
                ContextCompat.getDrawable(mActivity!!, iconRes),
                ContextCompat.getColor(mActivity!!, R.color.app_chrome_control_unchecked)
            )
        )
    }

    /** 音频卡片 + 播放/暂停、删除、信息三按钮的触摸 ripple 改为当前记事颜色（卡片随圆角、按钮圆形）。 */
    private fun applyAudioRipples(holder: AudioCardViewHolder) {
        val bg = mAccentBackground ?: ThingBackground.pure(mAccentColor)
        holder.cv!!.foreground =
            GradientRippleDrawable(bg, shapeOval = false, cornerRadiusPx = holder.cv!!.radius)
        // 三个按钮的 ripple 铺满按钮矩形区域（不是圆形）。
        holder.ivFirst!!.background = GradientRippleDrawable(bg, shapeOval = false, cornerRadiusPx = 0f)
        holder.ivSecond!!.background = GradientRippleDrawable(bg, shapeOval = false, cornerRadiusPx = 0f)
        holder.ivThird!!.background = GradientRippleDrawable(bg, shapeOval = false, cornerRadiusPx = 0f)
    }

    private fun startPlaying(index: Int) {
        mPlayingIndex = index
        val typePathName = mItems!![index]
        val file = File(typePathName!!.substring(1, typePathName.length))

        mPlayer = MediaPlayer.create(
            mActivity,
            FileProvider.getUriForFile(mActivity!!, "com.ywwynm.everythingdone", file)
        )
        mPlayer!!.setOnCompletionListener {
            val index1 = mPlayingIndex
            stopPlaying()
            notifyItemChanged(index1)
        }
        mPlayer!!.start()
    }

    open fun stopPlaying() {
        mPlayingIndex = -1
        mPlayer!!.stop()
        mPlayer!!.reset()
        mPlayer!!.release()
        mPlayer = null
    }

    inner class AudioCardViewHolder internal constructor(itemView: View?) : BaseViewHolder(itemView) {

        val cv: CardView?  = f(R.id.cv_audio_attachment)
        val tvName: TextView?   = f(R.id.tv_audio_file_name)
        val tvSize: TextView?   = f(R.id.tv_audio_size)
        val ivFirst: ImageView?  = f(R.id.iv_card_audio_first)
        val ivSecond: ImageView? = f(R.id.iv_card_audio_second)
        val ivThird: ImageView?  = f(R.id.iv_card_audio_third)

        init {
            setAudioIcon(ivThird, R.drawable.act_show_attachment_info)

            cv!!.setOnClickListener {
                togglePlay()
            }

            ivThird!!.setOnClickListener {
                val item = mItems!![this@AudioCardViewHolder.adapterPosition]
                val pathName = item!!.substring(1, item.length)
                AttachmentHelper.showAttachmentInfoDialog(mActivity, mAccentBackground, pathName)
            }

            if (mEditable) {
                setEventsEditable()
            } else {
                setEventsUneditable()
            }
        }

        private fun setEventsEditable() {
            ivFirst!!.setOnClickListener {
                togglePlay()
            }

            ivSecond!!.setOnClickListener {
                val pos = adapterPosition
                if (mPlayingIndex == pos) {
                    stopPlaying()
                    notifyItemChanged(pos)
                } else {
                    if (mRemoveCallback != null) {
                        mRemoveCallback!!.onRemoved(pos)
                    }
                }
            }
        }

        private fun togglePlay() {
            val pos = adapterPosition
            if (mPlayingIndex == pos) {
                if (mPlayer!!.isPlaying) {
                    mPlayer!!.pause()
                } else {
                    mPlayer!!.start()
                }
            } else {
                if (mPlayingIndex != -1) {
                    val index = mPlayingIndex
                    stopPlaying()
                    notifyItemChanged(index)
                }
                startPlaying(pos)
            }
            notifyItemChanged(pos)
        }

        private fun setEventsUneditable() {
            ivFirst!!.setOnClickListener {
                if (mPlayer!!.isPlaying) {
                    mPlayer!!.pause()
                } else {
                    mPlayer!!.start()
                }
                notifyItemChanged(adapterPosition)
            }

            ivSecond!!.setOnClickListener {
                val pos = adapterPosition
                if (mPlayingIndex == pos) {
                    stopPlaying()
                } else {
                    if (mPlayingIndex != -1) {
                        val index = mPlayingIndex
                        stopPlaying()
                        notifyItemChanged(index)
                    }
                    startPlaying(pos)
                }
                notifyItemChanged(pos)
            }
        }
    }

    companion object {
        const val TAG: String = "AudioAttachmentAdapter"
    }
}
